sycl/doc/design/IVDEPAndParallelLoopAnnotations.md

```diff
@@ -0,0 +1,455 @@
+guarantees about loop-carried memory dependence distance for loops with
+parallelism annotations like ivdep.
+
+## Attributes
+
+### Attributes for specifying no loop-carried dependences: `[[intel::ivdep]]` and `[[intel::ivdep(array)]]`
+
+When placed on a kernel loop, these attributes indicate that there are no
+loop-carried memory (data) dependences on all pointer and array accesses
+(`ivdep`) or on the specified array/pointer (`ivdep(array)`). This annotation
+allows the compiler to avoid making conservative assumptions when it cannot
+infer the loop-carried dependences on a specific array. For example consider the
+following code:
+
+```cpp
+for (int i = 0; i < N; ++i) {
+   B[i] += B[idx[i]];
+}
+```
+
+Since the access `B[idx[i]]` is unknown at compile time, the compiler will have
+to make the conservative assumption that there is a loop carried dependence of
+distance 1 between the write to B and the read from B (in both directions) if it
+tries to parallelize the loop. If however `ivdep` is used, then the annotation
+tells the compiler that there are no loop-carried dependences on `B`, and the
+compiler can make parallelism decisions without this assumption.
+
+```cpp
+[[intel::ivdep]]
+for (int i = 0; i < N; ++i) {
+   B[i] += B[idx[i]];
+}
+```
+
+### Attributes for specifying loop-carried dependence distance: `[[intel::ivdep(safelen)]]` and `[[intel::ivdep(array,safelen)]]`
+
+When placed on a loop, these attributes indicate that any loop-carried
+dependences on all pointer and array accesses (`ivdep(safelen)`) or on the
+specified array/pointer (`ivdep(array,safelen)`) have a distance of at least
+`safelen`. This can be used to disambiguate any aliasing when the compiler
+cannot automatically infer the dependence distance and instead has to make the
+conservative assumption that the distance is 1.
+
+In order to mark multiple arrays/pointers using ivdep on the same loop, then
+multiple instances of `ivdep(array)` or `ivdep(array,safelen)` can be used. The
+example below has the two arrays A and B marked with ivdep, but not C.
+Additionally, we indicate that the minimum dependence distance on accesses of B
+is 5 iterations.
+
+```cpp
+[[intel::ivdep(A)]]
+[[intel::ivdep(B,5)]]
+for(int i = 0; i <N; ++i) {
+   A[i] += A[idx_A[i]];
+   B[i] += B[idx_B[i]];
+   C[i] += C[idx_C[i]];
+}
+```
+
+### Marking parallel loops
+
+In contrast with `ivdep`, which only deals with memory dependences, a new
+language feature can be introduced to indicate that a loop is completely
+parallel (i.e., in addition to memory data dependences, any memory-ordering
+dependences can also be ignored). While we are not currently proposing this
+feature, we mention it for completeness and because there is an existing LLVM IR
+representation that has these semantics. If desired, it can be achieved using a
+new annotation `[[intel::parallel]]`, a new lambda-style `parallel_for` API, or
+using SYCL nested parallelism.
```

Comment:
Will FE add support for new attribute `[[intel::parallel]]'`?

Reply:
@smanna12 There is not current plan to add that as a new attribute. The design document mentions this feature for completeness and provides options in case we want to add this in the future.